File: agents/windows-winui/Services/DeviceMessageClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;

namespace LiveDashboardAgent.Services;

public sealed class DeviceMessageClient : IDisposable
{
    private static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(15);
    private readonly HttpClient _httpClient;
    private readonly string _serverUrl;
    private readonly string _token;

    public DeviceMessageClient(AppConfig config)
    {
        var normalized = config.Normalize();
        _serverUrl = normalized.ServerUrl.TrimEnd('/');
        _token = normalized.Token;
        _httpClient = new HttpClient
        {
            Timeout = DefaultTimeout,
        };
    }

    public async Task<IReadOnlyList<DeviceMessage>> FetchHistoryAsync(
        string since = "",
        CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, HistoryEndpoint(since));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);

        using var response = await _httpClient
            .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new DeviceMessageClientException(
                $"消息历史请求失败: HTTP {(int)response.StatusCode} {response.ReasonPhrase}".Trim(),
                (int)response.StatusCode,
                body);
        }

        return DeviceMessageResponseParser.Parse(body);
    }

    public async Task<IReadOnlyList<DeviceMessage>> FetchPendingAsync(
        CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(
            HttpMethod.Get,
            new Uri(_serverUrl + "/api/messages", UriKind.Absolute));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);

        using var response = await _httpClient
            .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new DeviceMessageClientException(
                $"待处理消息请求失败: HTTP {(int)response.StatusCode} {response.ReasonPhrase}".Trim(),
                (int)response.StatusCode,
                body);
        }

        return DeviceMessageResponseParser.Parse(body);
    }

    public async Task<bool> SendCommandAckAsync(
        IReadOnlyDictionary<string, object?> frame,
        CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            new Uri(_serverUrl + "/api/supervision/ack", UriKind.Absolute));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);
        request.Content = JsonContent.Create(frame);

        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        return response.IsSuccessStatusCode;
    }

    public void Dispose()
    {
        _httpClient.Dispose();
    }

    private Uri HistoryEndpoint(string since)
    {
        var suffix = "/api/messages/history";
        if (!string.IsNullOrWhiteSpace(since))
        {
            suffix += "?since=" + Uri.EscapeDataString(since);
        }
        return new Uri(_serverUrl + suffix, UriKind.Absolute);
    }
}

public sealed class DeviceMessageClientException : Exception
{
    public DeviceMessageClientException(string message, int statusCode, string responseBody)
        : base(message)
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
    }

    public int StatusCode { get; }
    public string ResponseBody { get; }
}

File: agents/windows-flutter/lib/src/services/device_message_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../models/app_config.dart';
import '../models/device_message.dart';

class DeviceMessageClient {
  DeviceMessageClient(AppConfig config)
    : _config = config.normalize(),
      _client = HttpClient() {
    _client.connectionTimeout = const Duration(seconds: 15);
  }

  static const _timeout = Duration(seconds: 15);
  final AppConfig _config;
  final HttpClient _client;

  Future<List<DeviceMessage>> fetchHistory({String since = ''}) {
    final suffix = since.isEmpty
        ? '/api/messages/history'
        : '/api/messages/history?since=${Uri.encodeQueryComponent(since)}';
    return _getMessages(suffix);
  }

  Future<List<DeviceMessage>> fetchPending() {
    return _getMessages('/api/messages');
  }

  Future<bool> sendCommandAck(Map<String, Object?> frame) async {
    try {
      final request = await _client
          .postUrl(_endpoint('/api/supervision/ack'))
          .timeout(_timeout);
      request.headers.contentType = ContentType.json;
      request.headers.set(
        HttpHeaders.authorizationHeader,
        'Bearer ${_config.token}',
      );
      request.write(jsonEncode(frame));
      final response = await request.close().timeout(_timeout);
      await response.drain<void>().timeout(_timeout);
      return response.statusCode >= 200 && response.statusCode < 300;
    } catch (_) {
      return false;
    }
  }

  Future<List<DeviceMessage>> _getMessages(String suffix) async {
    final request = await _client.getUrl(_endpoint(suffix)).timeout(_timeout);
    request.headers.set(
      HttpHeaders.authorizationHeader,
      'Bearer ${_config.token}',
    );
    final response = await request.close().timeout(_timeout);
    final body = await utf8.decodeStream(response).timeout(_timeout);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw DeviceMessageClientException(
        '消息请求失败: HTTP ${response.statusCode} ${response.reasonPhrase}'.trim(),
        response.statusCode,
        body,
      );
    }
    return DeviceMessage.listFromJson(body);
  }

  Uri _endpoint(String suffix) {
    return Uri.parse(
      '${_config.serverUrl.replaceFirst(RegExp(r'/+$'), '')}$suffix',
    );
  }

  void close() {
    _client.close(force: true);
  }
}

class DeviceMessageClientException implements Exception {
  const DeviceMessageClientException(
    this.message,
    this.statusCode,
    this.responseBody,
  );

  final String message;
  final int statusCode;
  final String responseBody;

  @override
  String toString() => message;
}
